src/solvers/pumpkin_solver.rs

```diff
@@ -0,0 +1,420 @@
+/// - Panic on ±∞ bounds
+///
+/// For the objective, we create a hidden `obj_var` that must match
+/// the user-provided linear expression, then do `maximise(obj_var)` or `minimise(obj_var)`.
+pub struct PumpkinProblem {
+    /// The underlying Pumpkin solver
+    solver: PumpkinSolver,
+    /// For each good_lp variable, which Pumpkin domain ID we use
+    domain_ids: Vec<DomainId>,
+
+    /// Whether we want to do max or min
+    direction: ObjectiveDirection,
+    /// The user’s entire objective expression
+    objective: Expression,
+
+    /// Next index for constraints
+    next_constraint_index: usize,
+    /// Prevent calling `solve()` multiple times
+    solved_once: bool,
+}
+
+impl PumpkinProblem {
+    /// Create a new Pumpkin problem from `UnsolvedProblem`.
+    fn new(to_solve: UnsolvedProblem) -> Self {
+        let UnsolvedProblem {
+            objective,
+            direction,
+            variables,
+        } = to_solve;
+
+        let mut solver = PumpkinSolver::default();
+        let mut domain_ids = Vec::with_capacity(variables.len());
+
+        // create an integer domain for each good_lp variable
+        for (_, def) in variables.iter_variables_with_def() {
+            // must be integer
+            if !def.is_integer {
+                panic!(
+                    "Pumpkin solver only supports integer variables, but got a non-integer definition: {:?}",
+                    def.name
+                );
+            }
+
+            // must have finite integral bounds
+            if def.min.is_infinite() || def.max.is_infinite() {
+                panic!(
+                    "Pumpkin solver does not handle infinite bounds for variable {:?}",
+                    def.name
+                );
+            }
+            let lb_i32 = i32_or_panic(def.min, "variable lower bound");
+            let ub_i32 = i32_or_panic(def.max, "variable upper bound");
+
+            let d_id = solver.new_bounded_integer(lb_i32, ub_i32);
+            domain_ids.push(d_id);
+        }
+
+        PumpkinProblem {
+            solver,
+            domain_ids,
+            direction,
+            objective,
+            next_constraint_index: 0,
+            solved_once: false,
+        }
+    }
+
+    /// Return an immutable reference to the underlying Pumpkin solver
+    pub fn as_inner(&self) -> &PumpkinSolver {
+        &self.solver
+    }
+
+    /// Return a mutable reference to the underlying Pumpkin solver
+    /// (use with caution, as changing the structure can break good_lp assumptions).
+    pub fn as_inner_mut(&mut self) -> &mut PumpkinSolver {
+        &mut self.solver
+    }
+}
+
+impl SolverModel for PumpkinProblem {
+    type Solution = PumpkinSolution;
+    type Error = ResolutionError;
+
+    fn solve(mut self) -> Result<Self::Solution, Self::Error> {
+        if self.solved_once {
+            panic!("PumpkinProblem: solve() called multiple times on same builder");
+        }
+        self.solved_once = true;
```

Comment:
what is the point of this ? Since you take ownership of self, this should not be needed? Or is there something I do not see ?

Reply:
arg sorry yes of course 🤦‍♂️ removing